Avoid recomputing deps unnecessarily

jqmp opened the pull request:
This is a substantial refactor of TaskCompletionRunner to avoid
recomputing non-persistable entities before loading their dependents
from the cache. The problem is this: in order to load a value from the
persistent cache, we need to compute the provenance digest of each of
its dependencies, which currently requires computing them. This means
that if a task has any non-persistable dependencies, they need to be
recomputed every time we want to load it from the cache. However, it's
also the case that when tasks have non-persistable output, their
provenance digest doesn't depend on their value, so this extra
computation is not even required.

To avoid this computation, we allow these non-persistable tasks to be in
a special "primed" state, where their digest has been computed but their
actual value has not. As it happens, this state is also useful for
tracking whether these tasks are ready to run in a subprocess.

I also made the following adjustments:

- I removed the DEFERRED stage for TaskEntries. I think the right way to
think about entry stages is that an entry is COMPLETED as soon as it
has met all known requirements; however, an entry can become
un-COMPLETED when new requirements become known.

- Previously task entries had a concept of "needed in memory" vs "needed
but not in memory". I changed this to "requested to have cached values"
vs "requested to be primed", which I think is a more useful distinction.

- The `_process_active_entry` method now contains the core of our
asynchronous logic, which has become more complex. The goal here is to
do as little work as possible to meet the requirements on an entry.

Fixes https://github.com/square/bionic/issues/111.

jqmp commented on bionic/core/task_execution.py:
```diff
@@ -16,46 +32,78 @@
+        # These are initially set to False, but can be updated to True when we discover
+        # a requirement from another entry. However, they should never go from True to
+        # False!
+        self.is_requested_to_be_primed = False
```
As discussed in a call, we should add a TODO to do the following:
1. Replace these two booleans with an enum.
2. Replace the `EntryBlockage` concept with some kind of `Requirement` object that represents a specific requirement level (using the enum above) on one entry from another.

namanjain commented on bionic/core/task_execution.py:
```diff
@@ -16,46 +32,78 @@
+        else:
+            return self.state.is_initialized
 
     def blocking_dep_entries(self):
```
Thanks for adding the other TODO. As we discussed on the same call, can you also leave a TODO here to merge the logic for `blocking_dep_entries` and `all_dependencies_satisfy_requests`?

namanjain commented on bionic/core/flow_execution.py:
```diff
@@ -108,32 +92,59 @@ def run(self, states):
+        for dep_entry in entry.dep_entries:
+            self._request_entry_be_primed(dep_entry)
+        if not entry.all_dependencies_satisfy_requests:
+            self._mark_entry_blocked(entry)
```
`_mark_entry_blocked` also tries to mark blocking entries pending, which `_request_entry_be_primed` and `_request_entry_be_cached` already do it for you. Can we combine those logic as well, maybe assert that blocking deps are not completed instead of marking them as pending?

jqmp replied:
Yes, good point -- thanks!

namanjain commented on bionic/core/flow_execution.py:
```diff
@@ -108,32 +92,59 @@ def run(self, states):
-        if state.is_cached:
+        # First, if we already have a cached value, we can stop immediately without
+        # needing to process any dependencies.
+        entry.state.set_up_caching_flags(self._bootstrap)
```
I don't think you need to set up the caching flags before initialize anymore. I had to set them up earlier because we used the caching flags to determine whether dependency entries were needed in memory.

namanjain commented on bionic/core/flow_execution.py:
```diff
@@ -145,28 +156,55 @@ def _process_entry(self, entry):
+
+        dep_entries = []
+        for dep_state in entry.state.dep_states:
+            dep_state.set_up_caching_flags(self._bootstrap)
```
Same here. I don't think we need this anymore.

jqmp commented:
I addressed the feedback and added two more changes:

1. I added a debug-level log statement for every entry stage-change. This makes debugging a lot easier!
2. I merged the `NEW` and `COMPLETED` stages. At this point I think they basically mean the same thing.

namanjain commented on bionic/core/task_execution.py:
```diff
@@ -147,65 +222,63 @@ def compute(self, task_key_logger):
 
     """
-    Entry is waiting to be processed.
+    The entry is waiting to be processed. We know there's work to be done on it, but
```
We create entries only when we need to process them. Since they start from `COMPLETED` state, we move them to `PENDING` to process them as soon as they are created.

How do you feel about making `PENDING` the starting stage? Or do you see a case in the future where we would create entries without knowing whether we need to process them? If not, I think it's better if we start from `PENDING` and make the change later on.

jqmp replied:
> We create entries only when we need to process them.

I'm not sure this is true. We create an entry when we need to express some requirement on it, but that requirement might already be met. If the task state for that entry is already primed or cached, we don't need to actually put the entry into a `PENDING` state.

More abstractly: When we create a new entry, it has no requirements at all, so it's "completed" by definition, right?

namanjain replied:
Duh, you are right. I was still thinking of it in terms of how we did it before. Also, you abstract explanation makes a lot of sense and convinced me easily!

namanjain commented on bionic/core/task_execution.py:
```diff
@@ -505,7 +571,7 @@ def set_up_caching_flags(self, bootstrap):
             To avoid this warning, disable persistence for the decorators
             by {disable_message}."""
-            logging.warn(message)
+            logger.warn(message)
```
This change does not belong in this PR but I just realized we are using `warnings.warn` in some places and `logger.warn` in other. We should make these warnings consistent.

namanjain commented on bionic/core/flow_execution.py:
```diff
@@ -210,7 +241,8 @@ def _mark_entry_blocked(self, blocked_entry):
 
-        self._mark_blocking_entries_pending(blocked_entry)
+        for blocking_entry in blocked_entry.blocking_dep_entries():
+            assert blocking_entry.stage != EntryStage.COMPLETED
```
Nice!

namanjain commented on bionic/core/task_execution.py:
```diff
@@ -585,17 +651,31 @@ def _load_value_hashes(self):
                     Failed to load cached value (hash) for descriptor
                     {accessor.query.dnode.to_descriptor()!r}.
                     This suggests we did not successfully complete the task
```
nit: `complete` -> `compute`

namanjain replied:
Same for line 364

jqmp commented:
It looks like the tests are failing because of a change in the behavior of Dask. We have a test that checks that we handle a particular exception it throws, but it seems to have stopped throwing that exception. I'll fix that in a separate PR before merging this one.
